src/main/java/com/rtb/projectmanagementtool/taskblocker/TaskBlockerController.java

```diff
@@ -25,58 +29,88 @@ public void addTaskBlocker(long taskID, long blockerID) {
   }
 
   public void addTaskBlocker(TaskBlockerData taskBlocker) {
-    // Get task
-    TaskData task;
-    try {
-      task = taskController.getTaskByID(taskBlocker.getTaskID());
-    } catch (NullPointerException e) {
-      System.out.println("Cannot find task with provided taskID");
+    // Ensure the tasks exist
+    ArrayList<TaskData> tasks =
+        taskController.getTasksByIDs(
+            new ArrayList<>(Arrays.asList(taskBlocker.getTaskID(), taskBlocker.getBlockerID())));
+    if (tasks.size() < 2) {
+      System.out.println("Cannot find tasks with provided taskID or blockerID");
       return;
     }
-    // Get blocker
-    TaskData blocker;
-    try {
-      blocker = taskController.getTaskByID(taskBlocker.getBlockerID());
-    } catch (NullPointerException e) {
-      System.out.println("Cannot find task with provided blockerID");
+    // Ensure the blocked task isn't already set to COMPLETE
+    if (tasks.get(0).getStatus() == Status.COMPLETE
+        || tasks.get(1).getStatus() == Status.COMPLETE) {
+      System.out.println("One or more tasks are already set to COMPLETE.");
       return;
     }
-    // Add taskBlocker
-    if (task.getStatus() != Status.COMPLETE) {
-      TransactionOptions options = TransactionOptions.Builder.withXG(true);
-      Transaction transaction = datastore.beginTransaction(options);
-      try {
-        datastore.put(taskBlocker.toEntity());
-        assert containsCycle() == false;
-        transaction.commit();
-      } finally {
-        if (transaction.isActive()) {
-          transaction.rollback();
-        }
-      }
-    } else {
-      System.out.println("Cannot block a task set to COMPLETE.");
+    // Ensure a cycle wouldn't be created if the TaskBlocker is added
+    if (containsPath(taskBlocker.getBlockerID(), taskBlocker.getTaskID())) {
+      System.out.println("Cannot block a task if it would create a cycle.");
+      return;
     }
+    // Add the TaskBlocker
+    taskBlocker.setTaskBlockerID(datastore.put(taskBlocker.toEntity()).getId());
   }
 
-  private boolean containsCycle() {
-    // TODO: Implement cycle detection algorithm
+  private boolean containsPath(long start, long end) {
```

Comment:
This would query the datastore + "build" and traverse the graph **every time** you need to add a new blocker.  (build in quotes because the algorithm you're using essentially builds and traverses the graph, just doesn't store the graph into a data structure).  Thoughts on optimizing?

Also, we don't have the feature at the moment, but if we build a "show all blockers of blockers" feature, optimization here could help with that as well.